public/calendar_reboot.js

```diff
@@ -865,6 +870,95 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
 };
 
 
+
+// Make a Firebase storage manager.
+var firebaseCalendarStorage = function(user) {
+    var self = this;
+    // Get a reference to the database service
+    self.database = firebase.database();
+    // This tells you whether the storage is actively working.
+    self.activeCalls = 0;
+    self.activityChangeFunctions = [];
+    
+    self.user = user;
+    
+    self.onActivityChanged = function(func) {
+        // Will run checkit's onActivityChanged.
+        
+        //TODO extend this to support multiple callbacks
+        self.activityChangeFunctions.push(func);
+    };
+    
+        self.startWork = function() {
+        // Will increment the counter and possibly fire an event.
+        
+        self.activeCalls += 1;
+        
+        // Will dispatch the event backgroundActivityChange 
+        self.activityChangeFunctions.forEach(function(func) {
+            func(self.activeCalls);
+        });
+        
+    };
+    
+    self.endWork = function() {
+        // Will decrement the counter and maybe fire an event.
+        
+        
+        self.activeCalls -= 1;
+        if (self.activeCalls < 0) {
+            console.error("No work has been started");
+        }
+        
+        // Dispatch the activityChanged listener
+        self.activityChangeFunctions.forEach(function(func) {
+            func(self.activeCalls);
+        });
+      
+    };
+    
+    self.getAllCalendarIds = function() {
+        // Returns a promise for the allCalendarIds object from storage
+        
+        var userId = self.user.currentUser.uid;
+        self.database.ref('/users/' + userId + '/checkit_allCalendarIds')
```

Comment:
This is the promise that you return. Call `startWork` before it starts and call `endWork` in the handlers instead of just logging.